File: testers/avro/java-vanilla/generated-and-checked-in/com/example/events/consumer/CustomerOrderConsumer.java
package com.example.events.consumer;

import com.example.events.CustomerOrder;
import com.example.events.header.StandardHeaders;
import java.time.Duration;
import org.apache.kafka.clients.consumer.Consumer;
import org.apache.kafka.clients.consumer.ConsumerRecords;

/** Type-safe consumer for customer-order topic */
public record CustomerOrderConsumer(
    Consumer<String, CustomerOrder> consumer, CustomerOrderHandler handler, String topic)
    implements AutoCloseable {
  public CustomerOrderConsumer(
      Consumer<String, CustomerOrder> consumer, CustomerOrderHandler handler) {
    this(consumer, handler, "customer-order");
  }

  public CustomerOrderConsumer withConsumer(Consumer<String, CustomerOrder> consumer) {
    return new CustomerOrderConsumer(consumer, handler, topic);
  }

  public CustomerOrderConsumer withHandler(CustomerOrderHandler handler) {
    return new CustomerOrderConsumer(consumer, handler, topic);
  }

  public CustomerOrderConsumer withTopic(String topic) {
    return new CustomerOrderConsumer(consumer, handler, topic);
  }

  /** Close the consumer */
  @Override
  public void close() {
    consumer.close();
  }

  /** Poll for messages and dispatch to handler */
  public void poll(Duration timeout) {
    ConsumerRecords<String, CustomerOrder> records = consumer.poll(timeout);
    records.forEach(
        record -> {
          String key = record.key();
          CustomerOrder value = record.value();
          StandardHeaders headers = StandardHeaders.fromHeaders(record.headers());
          handler.handle(key, value, headers);
        });
  }
}

File: testers/combined/kotlin/generated-and-checked-in/avro_events/combined/avro_events/consumer/CustomerOrderConsumer.kt
package combined.avro_events.consumer

import com.example.events.CustomerOrder
import combined.avro_events.header.StandardHeaders
import io.smallrye.mutiny.Multi
import io.smallrye.mutiny.Uni
import java.io.Closeable
import java.time.Duration
import org.apache.kafka.clients.consumer.Consumer
import org.apache.kafka.clients.consumer.ConsumerRecords

/** Type-safe consumer for customer-order topic */
data class CustomerOrderConsumer(
  val consumer: Consumer<kotlin.String, CustomerOrder>,
  val handler: CustomerOrderHandler,
  val topic: kotlin.String = "customer-order"
) : Closeable {
  /** Close the consumer */
  override fun close() {
    consumer.close()
  }

  /** Poll for messages and dispatch to handler, returning composed effect */
  fun poll(timeout: Duration): Uni<Unit> {
    val records: ConsumerRecords<kotlin.String, CustomerOrder> = consumer.poll(timeout)
    return Multi.createFrom().iterable(records).onItem().transformToUniAndConcatenate({ record -> handler.handle(record.key(), record.value(), StandardHeaders.fromHeaders(record.headers())) }).collect().last()
  }
}